### PR TITLE
WIP - IDR add option to connect to private omero instance

### PR DESCRIPTION
**Pull request by sunyi000**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/idr_download/idr_download_by_ids.xml [hunk 1]
```diff
@@ -39,10 +45,34 @@
...
 <inputs>
+        <conditional name="omero_instance_type">
+            <param name="omero_instance" type="select" label="Which OMERO instance to connect?"
+                   help="By default, the tool will download a tarball containing individual images from IDR into your Galaxy history. If you 
+                   need to connect to your own instance, set your connection username and password from User->Preference->Manage Information" >
+                <option value="idr">IDR</option>
+                <option value="priv">Your own instance</option>
+            </param>
+            <when value="priv">
+                <param name="omero_host" type="text" label="OMERO host URL">
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:67]: can we add here a validator?

File: tools/idr_download/idr_download_by_ids.xml [hunk 2]
```diff
@@ -39,10 +45,34 @@
...
+                </param>
+                <param name="omero_secure" type="boolean" label="Seured connection?" checked="true" truevalue="True" falsevalue="False"
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:69]: you can use truevalue="-s" here and falsevalue="" ...

File: tools/idr_download/idr_download_by_ids.py [hunk 3]
```diff
@@ -356,6 +363,18 @@ def _center_to_ul(center_x, center_y, width, height):
     p.add_argument(
         '--download-tar', action='store_true'
     )
+    p.add_argument(
+        '-h', '--omero-host', type=str, default="idr.openmicroscopy.org"
+    )
+    p.add_argument(
+        '-s', '--omero-secure', type=bool, default=True
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.py:370]: You can use here: `action='store_true'`

File: tools/idr_download/idr_download_by_ids.xml [hunk 4]
```diff
@@ -39,10 +45,34 @@
         #end if
         $skip_failed
         $download_tar
+        #if $omero_instance=='priv'':
+            -h $omero_host
+            -s $omero_secure
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:50]: then this becomes simply `-s` and with the truevalue trick below ... simply `$omero_secure`

File: tools/idr_download/idr_download_by_ids.xml [hunk 5]
```diff
@@ -20,6 +20,12 @@
         <requirement type="package" version="0.4.2">pylibtiff</requirement>
     </requirements>
     <command detect_errors="exit_code"><![CDATA[
+        
+        #if $omero_instance == "priv":
+            #set $username = $__user__.extra_preferences.get('omero_account|username', "")
+            #set $password = $__user__.extra_preferences.get('omero_account|password', "")
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:26]: What do you think, should we check here if both $username and $password are non-empty and provide a proper error message if they are empty? Something like ... you need to set those credentials in your user-prefeences. Please talk to your Galaxy administrator if you need further help?

**Review by bgruening — COMMENTED** [review]
* should we bump the omero dependency as well to 5.7.1?

Thanks @sunyi000

File: tools/idr_download/idr_download_by_ids.xml (hunk 5 above)
Comment (sunyi000) [reply tools/idr_download/idr_download_by_ids.xml:26]: Thanks for the quick review..I will address these one by one...

I had thought about this,, maybe just a 'help' text next to the "omero_instance " param is suffice?
I think it is an user error ift any or both are empty, with connect to private instance is selected?

 <param name="omero_instance" type="select" label="Which OMERO instance to connect?"
                   help="By default, the tool will download a tarball containing individual images from IDR into your Galaxy history. If you need to connect to your own instance, set your connection username and password from User->Preference->Manage Information">

File: tools/idr_download/idr_download_by_ids.xml (hunk 1 above)
Comment (sunyi000) [reply tools/idr_download/idr_download_by_ids.xml:67]: is it fine to just added an "empty_field" validator?

Comment (bgruening) [reply tools/idr_download/idr_download_by_ids.xml:67]: We could check for a URL with a regex and maybe only allow alphanumeric characters ...

File: tools/idr_download/idr_download_by_ids.xml (hunk 5 above)
Comment (bgruening) [reply tools/idr_download/idr_download_by_ids.xml:26]: I think this should go into the tool help at the bottom and also in a separate README file with instructions for the admins. The EGA repo I pointed you to has a nice example.

Comment (sunyi000) [reply tools/idr_download/idr_download_by_ids.xml:26]: README was already included. I will add the help text

File: tools/idr_download/idr_download_by_ids.xml (hunk 1 above)
Comment (sunyi000) [reply tools/idr_download/idr_download_by_ids.xml:67]: but in all the examples I can find, the BlitzGateway connection takes host as "idr.openmicroscopy.org" for example, without the http protocol.. so not sure if we should check for URL. But I can add validation for only allow alphanumeric characters and dots

**Comment by sunyi000** [comment]
@bgruening can you please have a look again? thanks

File: tools/idr_download/README.md [hunk 6]
```diff
@@ -0,0 +1,18 @@
+## Set up user credentials on Galaxy to connect to private omero instance
+
+To enable users to set their credentials for this tool,
+make sure the file `config/user_preferences_extra.yml` has the following section:
+
+```
+    omero_account:
+        description: Your OMERO instsnce connection credentials
```
Comment (bgruening) [inline tools/idr_download/README.md:8]: instance

File: tools/idr_download/idr_download_by_ids.xml [hunk 7]
```diff
@@ -16,10 +16,22 @@
...
+        
+        #if $omero_instance_type.omero_instance == "priv":
+            #set $username = $__user__.extra_preferences.get('omero_account|username', "")
+            #set $password = $__user__.extra_preferences.get('omero_account|password', "")
+
+            #if $username == "" or $password =="":
+                #set $username = "public"
+                #set $password = "public"
+            #end if
+            echo "OMERO connection credentials are empty. Set your credentials via: User -> Preferences -> Manage Information"  &&
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:32]: I would really error here out and not set username and password to public.

File: tools/idr_download/idr_download_by_ids.xml [hunk 8]
```diff
@@ -39,10 +51,35 @@
         #end if
         $skip_failed
         $download_tar
-
+        #if $omero_instance_type.omero_instance =='priv':
+            -h $omero_host
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:55]: single quotes

File: tools/idr_download/idr_download_by_ids.xml [hunk 9]
```diff
@@ -39,10 +51,35 @@
...
+            $omero_secured
+            -u $username
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:57]: single quotes

File: tools/idr_download/idr_download_by_ids.xml [hunk 10]
```diff
@@ -39,10 +51,35 @@
...
+        #end if
         2> >(tee -a $out_log >&2)
     ]]></command>
 <inputs>
+        <conditional name="omero_instance_type">
+            <param name="omero_instance" type="select" label="Which OMERO instance to connect?"
+                   help="By default, the tool will download a tarball containing individual images from IDR into your Galaxy history. If you 
+                   need to connect to your own instance, set your connection username and password from User->Preference->Manage Information" >
+                <option value="idr">IDR</option>
+                <option value="priv">Your own instance</option>
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.xml:68]: maybe we can name this `other Omero instance`

**Review by bgruening — COMMENTED** [review]
Sure, small review is inline, Thanks @sunyi000

**Comment by sunyi000** [comment]
> Sure, small review is inline, Thanks @sunyi000

Thanks @bgruening I made another round of changes based on previous review...
can you check the "username validation" part in the xml...not very sure if that's the correct way to error out..

Thanks

**Review by bgruening — APPROVED** [review]
LGTM, thanks @sunyi000!

**Pull request merged**